File: src/operations/system.py
"""
System operations handler for the auto-workflow system.
"""
import os
import subprocess
import time
import platform
from typing import Dict, Any, List
from .handler import Handler
from ..models import ExecutionContext


class SystemHandler(Handler):
    """Handles system operations."""

    def get_type(self) -> str:
        """Return the operation type."""
        return "system"

    def validate(self, params: Dict[str, Any]) -> None:
        """Check if the parameters are valid."""
        action = params.get("action")
        if not action or not isinstance(action, str):
            raise ValueError("action parameter is required")

        if action in ["command", "shell"]:
            if "command" not in params:
                raise ValueError(f"command parameter is required for {action} action")
        elif action == "sleep":
            if "duration" not in params:
                raise ValueError("duration parameter is required for sleep action")
        elif action == "open":
            if "path" not in params:
                raise ValueError("path parameter is required for open action")
        elif action == "env":
            if "variable" not in params:
                raise ValueError("variable parameter is required for env action")

    def execute(self, params: Dict[str, Any], exec_ctx: ExecutionContext) -> Dict[str, Any]:
        """Perform the system operation."""
        action = params["action"]

        if action == "command":
            return self._execute_command(params)
        elif action == "shell":
            return self._execute_shell(params)
        elif action == "sleep":
            return self._execute_sleep(params)
        elif action == "open":
            return self._execute_open(params)
        elif action == "env":
            return self._execute_env(params)
        elif action == "kill":
            return self._execute_kill(params)
        else:
            raise ValueError(f"unsupported system action: {action}")

    def _execute_command(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute command operation."""
        command = self._get_string_param(params, "command")
        working_dir = self._get_string_param(params, "working_dir")

        # Parse command and arguments
        parts = command.split()
        if not parts:
            raise ValueError("empty command")

        cmd = parts[0]
        args = parts[1:]

        # Set environment variables if provided
        env = os.environ.copy()
        if "env" in params and isinstance(params["env"], dict):
            for k, v in params["env"].items():
                env[str(k)] = str(v)

        start_time = time.time()
        result = subprocess.run(
            [cmd] + args,
            capture_output=True,
            text=True,
            cwd=working_dir if working_dir else None,
            env=env
        )
        duration = int((time.time() - start_time) * 1000)

        return {
            "success": result.returncode == 0,
            "command": command,
            "output": result.stdout,
            "error": result.stderr if result.returncode != 0 else None,
            "duration": duration,
        }

    def _execute_shell(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute shell command operation."""
        command = self._get_string_param(params, "command")
        working_dir = self._get_string_param(params, "working_dir")

        # Set environment variables if provided
        env = os.environ.copy()
        if "env" in params and isinstance(params["env"], dict):
            for k, v in params["env"].items():
                env[str(k)] = str(v)

        if working_dir:
            shell_command = f"cd {working_dir} && {command}"
        else:
            shell_command = command

        start_time = time.time()
        result = subprocess.run(
            shell_command,
            capture_output=True,
            text=True,
            shell=True,
            env=env
        )
        duration = int((time.time() - start_time) * 1000)

        return {
            "success": result.returncode == 0,
            "command": command,
            "output": result.stdout,
            "error": result.stderr if result.returncode != 0 else None,
            "duration": duration,
        }

    def _execute_sleep(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute sleep operation."""
        duration = self._get_int_param(params, "duration")

        time.sleep(duration)

        return {
            "success": True,
            "duration": duration,
        }

    def _execute_open(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute open operation."""
        path = self._get_string_param(params, "path")

        system = platform.system()

        if system == "Darwin":  # macOS
            if path.endswith(".app"):
                subprocess.run(["open", "-a", path], check=True)
            else:
                subprocess.run(["open", path], check=True)
        elif system == "Windows":
            os.startfile(path)
        else:  # Linux and others
            if "://" in path:
                subprocess.run(["xdg-open", path], check=True)
            else:
                subprocess.run(["xdg-open", path], check=True)

        return {
            "success": True,
            "path": path,
        }

    def _execute_env(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute env operation."""
        variable = self._get_string_param(params, "variable")
        value = self._get_string_param(params, "value", "")

        if value == "":
            # Get environment variable
            env_value = os.environ.get(variable, "")
            return {
                "success": True,
                "variable": variable,
                "value": env_value,
            }
        else:
            # Set environment variable (only for current process)
            os.environ[variable] = value
            return {
                "success": True,
                "variable": variable,
                "value": value,
            }

    def _execute_kill(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute kill operation."""
        process_name = self._get_string_param(params, "process_name")
        pid = self._get_int_param(params, "pid")

        if pid > 0:
            # Kill by PID
            try:
                os.kill(pid, 9)  # SIGKILL
            except ProcessLookupError:
                raise RuntimeError(f"process with PID {pid} not found")
        elif process_name:
            # Kill by process name
            system = platform.system()
            if system == "Darwin" or system == "Linux":
                subprocess.run(["pkill", process_name], check=True)
            elif system == "Windows":
                subprocess.run(["taskkill", "/F", "/IM", process_name], check=True)
        else:
            raise ValueError("either process_name or pid must be specified")

        return {
            "success": True,
        }

    # Helper methods
    def _get_string_param(self, params: Dict[str, Any], key: str, default: str = "") -> str:
        """Get string parameter."""
        value = params.get(key)
        if isinstance(value, str):
            return value
        return str(value) if value is not None else default

    def _get_int_param(self, params: Dict[str, Any], key: str, default: int = 0) -> int:
        """Get integer parameter."""
        value = params.get(key)
        if value is None:
            return default
        if isinstance(value, int):
            return value
        if isinstance(value, float):
            return int(value)
        if isinstance(value, str):
            try:
                return int(value)
            except ValueError:
                return default
        return default

File: pkg/operations/system.go
package operations

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"strings"
	"time"
)

// SystemHandler handles system operations
type SystemHandler struct{}

// NewSystemHandler creates a new system handler
func NewSystemHandler() *SystemHandler {
	return &SystemHandler{}
}

// GetType returns the operation type
func (h *SystemHandler) GetType() string {
	return "system"
}

// Validate checks if the parameters are valid
func (h *SystemHandler) Validate(params map[string]interface{}) error {
	action, ok := params["action"].(string)
	if !ok {
		return fmt.Errorf("action parameter is required")
	}

	switch action {
	case "command", "shell":
		if _, ok := params["command"]; !ok {
			return fmt.Errorf("command parameter is required for %s action", action)
		}
	case "sleep":
		if _, ok := params["duration"]; !ok {
			return fmt.Errorf("duration parameter is required for sleep action")
		}
	case "open":
		if _, ok := params["path"]; !ok {
			return fmt.Errorf("path parameter is required for open action")
		}
	case "env":
		if _, ok := params["variable"]; !ok {
			return fmt.Errorf("variable parameter is required for env action")
		}
	}

	return nil
}

// Execute performs the system operation
func (h *SystemHandler) Execute(ctx context.Context, params map[string]interface{}, execCtx interface{}) (map[string]interface{}, error) {
	action := params["action"].(string)

	var result map[string]interface{}
	var err error

	switch action {
	case "command":
		result, err = h.executeCommand(ctx, params)
	case "shell":
		result, err = h.executeShell(ctx, params)
	case "sleep":
		result, err = h.executeSleep(ctx, params)
	case "open":
		result, err = h.executeOpen(ctx, params)
	case "env":
		result, err = h.executeEnv(ctx, params)
	case "kill":
		result, err = h.executeKill(ctx, params)
	default:
		return nil, fmt.Errorf("unsupported system action: %s", action)
	}

	return result, err
}

func (h *SystemHandler) executeCommand(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	command := getStringParam(params, "command")
	workingDir := getStringParam(params, "working_dir")

	// Parse command and arguments
	parts := strings.Fields(command)
	if len(parts) == 0 {
		return nil, fmt.Errorf("empty command")
	}

	cmd := exec.CommandContext(ctx, parts[0], parts[1:]...)

	if workingDir != "" {
		cmd.Dir = workingDir
	}

	// Set environment variables if provided
	if env, ok := params["env"].(map[string]interface{}); ok {
		for k, v := range env {
			cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%v", k, v))
		}
	}

	startTime := time.Now()
	output, err := cmd.CombinedOutput()
	duration := time.Since(startTime)

	if err != nil {
		return map[string]interface{}{
			"success":  false,
			"command":  command,
			"output":   string(output),
			"error":    err.Error(),
			"duration": duration.Milliseconds(),
		}, nil
	}

	return map[string]interface{}{
		"success":  true,
		"command":  command,
		"output":   string(output),
		"duration": duration.Milliseconds(),
	}, nil
}

func (h *SystemHandler) executeShell(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	command := getStringParam(params, "command")
	workingDir := getStringParam(params, "working_dir")

	var cmd *exec.Cmd
	if workingDir != "" {
		cmd = exec.CommandContext(ctx, "sh", "-c", "cd "+workingDir+" && "+command)
	} else {
		cmd = exec.CommandContext(ctx, "sh", "-c", command)
	}

	// Set environment variables if provided
	if env, ok := params["env"].(map[string]interface{}); ok {
		for k, v := range env {
			cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%v", k, v))
		}
	}

	startTime := time.Now()
	output, err := cmd.CombinedOutput()
	duration := time.Since(startTime)

	if err != nil {
		return map[string]interface{}{
			"success":  false,
			"command":  command,
			"output":   string(output),
			"error":    err.Error(),
			"duration": duration.Milliseconds(),
		}, nil
	}

	return map[string]interface{}{
		"success":  true,
		"command":  command,
		"output":   string(output),
		"duration": duration.Milliseconds(),
	}, nil
}

func (h *SystemHandler) executeSleep(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	duration := getIntParam(params, "duration")

	select {
	case <-time.After(time.Duration(duration) * time.Second):
		return map[string]interface{}{
			"success":  true,
			"duration": duration,
		}, nil
	case <-ctx.Done():
		return nil, ctx.Err()
	}
}

func (h *SystemHandler) executeOpen(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	path := getStringParam(params, "path")

	var cmd *exec.Cmd
	switch {
	case strings.HasSuffix(path, ".app"):
		// macOS application
		cmd = exec.Command("open", "-a", path)
	case strings.Contains(path, "://"):
		// URL
		cmd = exec.Command("open", path)
	default:
		// File or directory
		cmd = exec.Command("open", path)
	}

	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("open failed: %v, output: %s", err, string(output))
	}

	return map[string]interface{}{
		"success": true,
		"path":    path,
	}, nil
}

func (h *SystemHandler) executeEnv(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	variable := getStringParam(params, "variable")
	value := getStringParam(params, "value")

	if value == "" {
		// Get environment variable
		envValue := os.Getenv(variable)
		return map[string]interface{}{
			"success":  true,
			"variable": variable,
			"value":    envValue,
		}, nil
	} else {
		// Set environment variable (only for current process)
		os.Setenv(variable, value)
		return map[string]interface{}{
			"success":  true,
			"variable": variable,
			"value":    value,
		}, nil
	}
}

func (h *SystemHandler) executeKill(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	processName := getStringParam(params, "process_name")
	pid := getIntParam(params, "pid")

	var err error
	if pid > 0 {
		// Kill by PID
		process, errFind := os.FindProcess(pid)
		if errFind != nil {
			return nil, fmt.Errorf("find process failed: %v", errFind)
		}
		err = process.Kill()
	} else if processName != "" {
		// Kill by process name (macOS/Linux)
		cmd := exec.Command("pkill", processName)
		output, errKill := cmd.CombinedOutput()
		if errKill != nil {
			return nil, fmt.Errorf("kill process failed: %v, output: %s", errKill, string(output))
		}
	} else {
		return nil, fmt.Errorf("either process_name or pid must be specified")
	}

	if err != nil {
		return nil, fmt.Errorf("kill failed: %v", err)
	}

	return map[string]interface{}{
		"success": true,
	}, nil
}
